overlord/servicestate/quota_handlers.go

```diff
@@ -560,6 +560,58 @@ type ensureSnapServicesForGroupOptions struct {
 	extraSnaps []string
 }
 
+func filterSnapServices(opts *wrappers.SnapServicesOptions, include []string) *wrappers.SnapServicesOptions {
+	if len(include) == 0 {
+		return opts
+	}
+
+	filteredMap := make(map[*snap.AppInfo]*quota.Group)
+	for svc, grp := range opts.ServiceQuotaMap {
+		if strutil.ListContains(include, svc.Name) {
+			filteredMap[svc] = grp
+		}
+	}
+	opts.ServiceQuotaMap = filteredMap
+	return opts
+}
+
+// affectedSnapServices returns a map of snaps and the services affected
+// by performing changes to a quota group. For groups that contain just snaps, all
+// services are added to the map, for groups that contain specific services, only those
+// services are affected.
+func affectedSnapServices(st *state.State, grp *quota.Group, opts *ensureSnapServicesForGroupOptions) (map[*snap.Info]*wrappers.SnapServicesOptions, error) {
+
+	// handle extra snaps also passed here, so that they get included
+	// in any case
+	snapServices := make(map[string][]string)
+	for _, sn := range append(grp.Snaps, opts.extraSnaps...) {
+		snapServices[sn] = []string{}
+	}
+
+	// if the group is a service group, then we need to get the affected
+	// snaps from the service names, which are of format 'snap.service'
+	for _, sn := range grp.Services {
+		parts := strings.Split(sn, ".")
+		snapName := parts[0]
+		svcName := parts[1]
+		snapServices[snapName] = append(snapServices[snapName], svcName)
+	}
+
+	snapSvcMap := map[*snap.Info]*wrappers.SnapServicesOptions{}
+	for sn, svcs := range snapServices {
+		info, err := snapstate.CurrentInfo(st, sn)
+		if err != nil {
+			return nil, err
+		}
+		opts, err := SnapServicesOptions(st, info, opts.allGrps)
+		if err != nil {
+			return nil, err
+		}
+		snapSvcMap[info] = filterSnapServices(opts, svcs)
```

Comment:
I'm not sure I understand this filtering because ultimately the code in wrappers.go  EnsureSnapServices -> ensureSnapsSystemdServices -> ensureSnapServiceSystemdUnits operates on s.Services() that means all the services anyway, so with this filtering we actually don't give it all the info?

Reply:
The code is correct, imagine you have a snap like this:

my-snap:
  services:
     - svc1
     - svc2
     
You put my-snap into quota group foo, you then create a sub-group called foo-sub, then put my-snap.svc1 into foo-sub.

Now you change the memory limit on foo-sub, which means we need to update the slice for foo-sub. Which services are affected? Only my-snap.svc1.

However, `SnapServicesOptions` will return both my-snap.svc1 and my-snap.svc2. So we need to filter out any services that are not listed in foo-sub.

Reply:
I understand what the code is doing and the intention. I question what really happens given the how things are implemented in EnsureSnapServices. This is not about the local behavior of this code, but about the ultimate effect of it.

Reply:
Ah you are absolutely correct! The changes we did to how `EnsureSnapServices` broke this approach. It just dawned on me, sorry, I misunderstood what you were trying to communicate! Because EnsureSnapServices now works on the entire list, this approach does not work anymore.